### PR TITLE
Oracle contract added

### PR DESCRIPTION
**Pull request by JaisPiyush**
- Oracle contract ownership management added
- Ability to register up to 3 oracles (for owner) is added
- Consensus variables and getters added
- `submit_sentinelData` added
- All of the above functions are tested

### CODE REVIEW

File: foundry.toml [hunk 1]
```diff
@@ -1,6 +1,24 @@
...
 src = 'src'
 out = 'out'
+test = 'test'
 libs = ['lib']
 
+remappings = [
+    'ds-test/=lib/forge-std/lib/ds-test/src/',
+    'forge-std/=lib/forge-std/src/',
+    'openzeppelin-contracts/=lib/openzeppelin-contracts/'
+]
```
Comment (hcheng826) [inline foundry.toml]: let's remove the `remapping.txt`?

File: src/Oracle.sol [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+
+    mapping(bytes32 => uint256) public dataFeedVotes;
+    mapping(bytes32 => bool) public nodeSubmissions;
+    mapping(bytes32 => bool) public executedSubmissions;
+
+    constructor() Ownable() {}
+
+    struct SentinelData {
+       uint256 height;
```
Comment (hcheng826) [inline src/Oracle.sol:38]: indent

File: src/Oracle.sol [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    event SentinelDataSubmitted(address indexed oracle, uint256 indexed height, uint256 indexed currentHeight);
+    event SentinelDataUpdated(uint256 indexed height, uint256 indexed currentHeight);
+
+    event OracleAdded(address indexed oracle, uint256 indexed oracleCount);
+
+    EnumerableSet.AddressSet private oracles;
+
+    // mapping(address => bool) public isRegisterdOracle;
+    // uint256 public totalRegisterdOracles;
+    uint8 public quorum = 51;
```
Comment (hcheng826) [inline src/Oracle.sol]: mark it constant to save gas?

File: src/Oracle.sol [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    /// Register oracle on contract owner's command
+    /// Maximum 3 oracles can be added by owners to seed the protocol
+    function registerOracleThroughOwner(address oracle) external onlyOwner {
+        require(totalRegisterdOracles() < 3, "Privilege exhausted");
+        _registerOracle(oracle);
+    }
+
+    function submit_SentinelData(
```
Comment (hcheng826) [inline src/Oracle.sol]: should we make this camel case?

File: src/Oracle.sol [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      uint256 height,
+      uint256 initialPledgeRateFor32GiB,
+      uint256 rawBytesPowerOfProtocol,
+      uint256 availableBalanceOfProtocol,
+      uint256 pledgedCollateralOfProtocol
```
Comment (hcheng826) [inline src/Oracle.sol]: consider just pass the entire struct as param?
```
function submiteSentinelData(SentinelData calldata sentinalData) {}
```

**Review by hcheng826 — COMMENTED** [review]
Is there a particular reason that we're not using the Ownable from Openzeppelin? Saw the `addOwner` function but it's not really used?

**Comment by JaisPiyush** [comment]
> Is there a particular reason that we're not using the Ownable from Openzeppelin? Saw the `addOwner` function but it's not really used?

It's inspired from the Openzepp. I just wrote it from scratch. addOwner is an external call to add another owner of the contract.

This contract stands between Single owner and role based owner Ownable contract in openzepp. It can have multiple owner but all will have same powers, just for sake of ease.

File: src/Oracle.sol (hunk 4 above)
Comment (JaisPiyush) [reply src/Oracle.sol]: Well I've used the same cade in all oracles abi, I can change this to pure camelcase

**Comment by hcheng826** [comment]
lgtm. do you want to change the base branch to main? or merge to this one `hcheng826:feature/oracle`? Feel free to merge it yourself after u change the base branch

**Review by hcheng826 — APPROVED** [review]

File: src/LoanAgentFactory.sol [hunk 6]
```diff
@@ -1,126 +1,244 @@
...
+        require(Utils.isOwnerOfTheMiner(request.miner, msg.sender), "Only owner can request for loan");
+        require(request.nodeOwnerStakePercentage >= 50 && request.nodeOwnerStakePercentage <= 75, "Owner stake out of bound");
+        require(request.timeCommitment >= MIN_TIME_COMMIT && 
+            request.timeCommitment <= MAX_TIME_COMMIT, "Time commitment out of bound");
+        require(request.oracle != address(0), "Invalid oracle address");
+        require(Utils.isEligibleRawBytePower(request.rawBytesPower), "Min. 3PiB storage");
+        uint256 expectedCollateral = Oracle(oracle).getIntialCollateralPrice(request.rawBytesPower) * request.nodeOwnerStakePercentage / ONE_HUNDRED_DENOMINATOR;
+        // 10% extra collateral
+        uint256 requiredPledge = expectedCollateral + (expectedCollateral * 10 / ONE_HUNDRED_DENOMINATOR);
+        require(msg.sender >= requiredPledge, "Value sent is less");
```
Comment (hcheng826) [inline src/LoanAgentFactory.sol:119]: should be `msg.value`

File: src/LoanAgentFactory.sol [hunk 7]
```diff
@@ -1,126 +1,244 @@
...
+        minerDataInQueue[agent].timestamp = block.timestamp;
+        LoanAgent(agent).updateNodeStatus(LoanAgent.NodeStatus.WaitingForOracleVerification);
+        Oracle(oracle).verifyOracleForRegistration(qData.processData.oracle);
+    }
+
+    function _initSectorPledging(address agent) internal {
+        // Transfer pledge amount and user stakes
+    }
+
+    function acceptMinerForLoan(address agent, uint256 currentInitialCollateral) external onlyOracle {
```
Comment (hcheng826) [inline src/LoanAgentFactory.sol:199]: will there be FIL transfer in this function? From `aFIL` to `LoanAgent`?

**Review by hcheng826 — COMMENTED** [review]
lgtm

**Pull request merged**